### ISSUE TITLE
feature: Kotlin like extension functions

### ISSUE THREAD

**Issue by MikeMitterer** [dart-lang/sdk#32280]
I know, I know - not possible for Dart 2 but maybe for Dart 2.1???

Extension functions are really!!!! cool in Kotlin - would be cool for Dart as well 
https://kotlinlang.org/docs/reference/extensions.html

**Comment by lrhn** [comment]
The Kotlin extension functions are a version of scoped and statically-resolved extension methods, similar to the ones in C#. That makes them easy to implement, but they're also just a type-based shorthand for calling a *static* method. It allows someone else to extend a type, but it doesn't allow a type to specialize the implementation. It also allows conflicts between different extensions and between an extension and the type itself. 

Another option is to define the extension methods on new "static types" that adapt the original types, so instead of adding `List<T>.rotate(int n)`, you introduce a new type `MyList<T>` that "extends" `List<T>` and has a `rotate(int n)` method. Then you do `MyList<T> myList = list; myList.rotate(42);`.
That will allow you to choose to adapt any list, and avoid conflicts between different extensions.
(It's list traits, basically).

There are multiple other options, all with different trade-offs, because nothing is ever easy :)

**Comment by svenjacobs** [comment]
@lrhn Yes, Kotlin's extension functions are "just" syntactic sugar and there are several other options in Dart (and other languages in general) like you described. However I think the point is **that** they are syntactic sugar and that they can be called like functions on the original class. This enables some nice, concise syntax. I, too, would like to see the support for extension functions (and properties?) in a future Dart version :)

**Comment by passsy** [comment]
I really want extension functions in dart. Especially the `Iterable` API is lacking some features and it would be super handy to be able to add them myself and even publish them as package.

### Example 1

Wrapping is not bad but sometimes breaks patterns.

With extension function
```dart
// with proposed extension functions
// distinctBy not in sdk, but works nicely with extension function
list.where((it) => it.size > 4)
    .map((it) => it.id)
    .distinctBy((it) => it)
    .skipWhile((it) => isValidId(it))

List<T> List<T>.distinctBy<T>(bool selector(T t)) => /*...*/;
```

Current solution
```dart
// current state of the art
// without extension and wrapping, therefore unreadable
distinctBy(list.where((it) => it.size > 4)
    .map((it) => it.id), (it) => it)
    .skipWhile((it) => isValidId(it));

List<T> distinctBy<T>(List<T> l, bool selector(T t)) => /*...*/;
```

### Example 2
Problem:
```dart
void doSomething(List<String> list) {
  // crashing for empty list with "No element"
  // firstOrNull not in sdk
  if (list.first == "abc") {
    print("found");
  }
}
```

Solution with extension function
```dart
// proposed extension function
T List<T>.firstOrNull<T> => this.isEmpty ? null : this.first;

void doSomething(List<String> list) {
  // nearly unchanged code
  if (list.firstOrNull == "abc") {
    print("found");
  }
}
```

Current solution:
```dart
// normal helper function, requires wrapping
T firstOrNull<T>(List<T> l) => l.isEmpty ? null : l.first;

void doSomething(List<String> list) {
  // fixed but list is now wrapped
  if (firstOrNull(list) == "abc") {
    print("found");
  }
}
```

**Comment by MikeMitterer** [comment]
The more I use extension functions the more I come to the conclusion that they are fare from "just syntactic sugar" - they are one of Kotlins "killer features"!

**Comment by loic-sharma** [comment]
Also, extensions can add common functionality to ALL children of a certain type. Extensions methods are also wonderful for composition. For example, an app can compose extensions to `List` from two different package dependencies. This doesn't work nearly as well if everyone is extending `List` with their own custom functionality.

**Comment by marcguilera** [comment]
It might "just" be sugar but it would be awesome. I find myself extending some classes like Observable just to add some functionality when I would like to declare the extension and be able to use it with the original instance elsewhere in the codebase. Is this on the radar at all?

**Comment by mpiannucci** [comment]
Just want to chime in that this would be incredibly useful for Flutter development.

**Comment by lrhn** [comment]
Closing this as a duplicate of #9991.

**Comment by erluxman** [comment]
If **Fast Development** is Flutter's USP and Dart cares for Flutter, then I think community deserves to get what they are asking for  `Extension Function in Dart`.

**Comment by dgrove** [comment]
https://github.com/dart-lang/language/issues/41 is where this is being
worked on.

**Comment by BaraaSoft** [comment]
I agree the extension method is wonderful, it must be added to Dart.

**Comment by eernstg** [comment]
Cf. dart-lang/language#41, dart-lang/language#42, dart-lang/language#177 for current discussions on this topic.

**Comment by derolf** [comment]
I am working on a project where the backend is written in Kotlin and the frontend in Dart (Flutter), so I am switching back and forth the whole day.

Extension functions and null-pointer-safety are definitely the number 1 features I miss!

**Comment by pedromassango** [comment]
Please Dart team add it. We need this

**Comment by NanoMichael** [comment]
That be very attractive to add this feature.

**Comment by RyanSusana** [comment]
Let's adddddd this!

**Comment by levirgon** [comment]
Add this, please. :+1:

**Comment by creativecreatorormaybenot** [comment]
@levirgon It is available [starting with Dart `2.6`](https://stackoverflow.com/q/58288235/6509751).
